File: packages/typescriptlib/src/main/typescript/service.ts
import {ServiceConfiguration} from './serviceConfiguration.js';

/**
 * A Service is a predefined component of functionality with a lifecycle
 * managed by the ServiceManger.
 */
interface Service<C extends ServiceConfiguration = ServiceConfiguration>
{
    /**
     * Initializes the service instance
     *
     * @param configuration Service level configuration
     */
    init(configuration: C): void;

    /**
     * Starts the service instances
     */
    start(): void;

    /**
     * Stops the service instance
     */
    stop(): void;

    /**
     * Destroys the service instance.  Can be used to clean up memory or other
     * resources
     */
    destroy(): void;
}

/**
 * Base instance of a Service which is a noop for all methods.  Extend to create
 * a new service and implement the methods that are required
 */
abstract class BaseService<C extends ServiceConfiguration = ServiceConfiguration> implements Service<C>
{
    destroy(): void
    {
        // Do Nothing Implementation
    }

    init(configuration: C): void
    {
        // Do Nothing Implementation
    }

    start(): void
    {
        // Do Nothing Implementation
    }

    stop(): void
    {
        // Do Nothing Implementation
    }
}

export {BaseService};
export type {Service};

File: packages/pythonlib/src/main/python/service_manager/service.py
from typing import TypeVar
from abc import ABC, abstractmethod
from service_configuration import ServiceConfiguration

CConfiguration = TypeVar("CConfiguration", bound="ServiceConfiguration")
TService = TypeVar("TService", bound="Service")

class Service[CConfiguration](ABC):
    """A Service is a predefined component of functionality with a lifecycle managed by the ServiceManger."""
    @abstractmethod
    def init(self, configuration: C) -> None:
        """Initialize the service instance

        Args:
            configuration (C): Service level configuration"""
        pass

    @abstractmethod
    def start(self) -> None:
        """Start the service instances"""
        pass

    @abstractmethod
    def stop(self) -> None:
        """Stop the service instance"""
        pass

    @abstractmethod
    def destroy(self) -> None:
        """Destroy the service instance.  Can be used to clean up memory or other resources"""
        pass


class BaseService[CConfiguration](Service):
    """Base instance of a Service which is a noop for all methods.  Extend to create a new service and implement the methods that are required"""

    def init(self, configuration: C) -> None:
        pass

    def start(self) -> None:
        pass

    def stop(self) -> None:
        pass

    def destroy(self) -> None:
        pass

class ServiceException(Exception):
    """
    Service API Exception
    """
    pass
